File: python/34.py

def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)

factorial_cache = {i: factorial(i) for i in range(10)}


def is_digit_factorial(n):
    if n in [1, 2]:
        return False
    return n == sum(map(lambda x: factorial_cache[int(x)], list(str(n))))

print(sum((i for i in range(2, 8 * factorial_cache[9] + 1) if is_digit_factorial(i))))

File: cpp/34.cpp
#include <iostream>
#include <vector>

using namespace std;


vector<int> factorial_cache {};

int factorial(int n){
    if (n == 0)
        return 1;
    else
        return n * factorial(n - 1);
}

bool is_digit_factorial(int n){
    if (n <= 2)
        return false;

    int s {};
    for (char c: to_string(n))
        s += factorial_cache[c - '0'];
    
    if (s == n)
        return true;
    else
        return false;
}


int main() {

    for (int i = 0; i < 10; i++)
    {
        factorial_cache.push_back(factorial(i));
    }
    
    int s {};

    for (int i = 2; i < 8 * factorial_cache[9] + 1; i++)
    {
        if (is_digit_factorial(i))
            s += i;
    }
    


    cout << s << endl;
    return 0;
}